README.rst
collective.firehose
===================

The ZODB is great at storing document-oriented data. But it's pretty bad
at heavy-write scenarios, such as counting hits to create a list of most
popular pages.

``collective.firehose`` makes this sort of accounting possible by using
`0mq`_ to relay request start and end events to a separate process which
records hits in `redis`_.

(0mq is a brokerless message queue, which keeps needed infrastructure
to a minimum. Redis is a data structure server which is great at storing
the kind of data we need for this use case. Both are fast and this
architecture keeps overhead on the Zope instance tiny.)

.. _`0mq`: http://www.zeromq.org/
.. _`redis`: http://redis.io/

Installation
------------

You need libzmq and redis, with a running redis server.

E.g., pre-installation on OS X using homebrew::

  $ brew install redis zmq
  $ redis-server /usr/local/etc/redis.conf

In the buildout for your Plone site, you need to add collective.firehose
to your instance eggs::

  [instance]
  eggs = collective.firehose

And you also need to install the script which will relay request info to
redis::

  [buildout]
  parts += firehose

  [firehose]
  recipe = zc.recipe.egg
  eggs = collective.firehose

After running buildout, you should have a bin/firehose-record script. Run
this.

Finally, fire up your instance and go to http://localhost:8080/firehose-stats
You should see the stats update every 5 seconds as you navigate around the
site in another window.

If you are collecting stats from multiple Zope instances, you can make it
easier to identify which instance is serving a particular request by giving
each instance a unique identifier. Add to your buildout::

  [instance]
  zope-conf-additional =
    <product-config firehose>
        instance_id my-instance-id
    </product-config>

Retrieving statistics
---------------------

Stats can be retrieved using the redis client library. For example, here's
how the included stats view gets the top 10 most popular URLs for the
current hour::

  import redis
  import time

  # Get redis connection
  r = redis.StrictRedis(host='localhost', port=6379, db=0)

  # Calculate redis key for current hour
  timeslot = time.time() // 3600

  # Get the first 10 members of the "tophits" sorted set for the current hour.
  # ``tophits`` will be a list of (url, hit_count) tuples.
  tophits = r.zrevrange('tophits.%s' % timeslot, 0, 10, withscores=True)

Firehose stores in the following keys in redis:

  ``serving``
    A set of the URLs currently being served by Zope backends.

  ``tophits.[hour]``
    A sorted set of URLs accessed in the specified hour, scored by # of hits.
    The hour is given as the current Unix timestamp // 3600.

  ``elapsed.[hour]``
    A sorted set of URLs accessed in the specified hour, scored by the length
    of the slowest request for that URL.

Summarizing stats by periods of time larger than an hour should be possible
by periodically merging the hourly buckets. This is left as an exercise for the
pull-requester. ;-)


CHANGES.txt
Changelog
=========

1.0 (unreleased)
----------------

* Initial release.
  []


sub.py
import time
import redis
import zmq


context = zmq.Context()
sub = context.socket(zmq.SUB)
sub.connect('ipc:///tmp/collective.firehose.sock')
sub.setsockopt(zmq.SUBSCRIBE, '')

r = redis.StrictRedis(host='localhost', port=6379, db=0)


while True:
    url, elapsed = sub.recv().rsplit(' ', 1)
    print url, elapsed
    pipe = r.pipeline()

    # track current requests
    if elapsed == '0':
    	pipe.sadd('serving', url)
    else:
    	pipe.srem('serving', url)

    # track top hits per hour
    timeslot = time.time() // 3600
    pipe.zincrby('tophits.%s' % timeslot, url, 1)

    # track slowest hits per hour (in ms)
    slowkey = 'elapsed.%s' % timeslot
    old_elapsed = r.zscore(slowkey, url)
    if old_elapsed is None or float(elapsed) > float(old_elapsed):
    	pipe.zincrby(slowkey, url, elapsed)

    pipe.execute()

# What data do I want to fetch?
# 1. What URL is each client serving right now -- set of client:URL
# 2. What are the most popular URLs? -- sorted set of URL scored by weight
# 3. What are the most popular URLs within the past hour? -- sorted set of URLs scored by weight for each hour + cron job that updates a union of the past N buckets
# 4. What are the slowest URLs? -- sorted set of URLs scored by request time


collective/firehose/recorder.py
import time
import redis
import zmq


def record_stats():

    try:
        context = zmq.Context()
        sub = context.socket(zmq.SUB)
        sub.bind('ipc:///tmp/collective.firehose.sock')
        sub.setsockopt(zmq.SUBSCRIBE, '')

        r = redis.StrictRedis(host='localhost', port=6379, db=0)

        while True:
            instance_id, rest = sub.recv().split(' ', 1)
            url, elapsed = rest.rsplit(' ', 1)
            pipe = r.pipeline()

            # track current requests
            if elapsed == '0':
                pipe.sadd('serving', '%s:%s' % (instance_id, url))
            else:
                pipe.srem('serving', '%s:%s' % (instance_id, url))

                # track top hits per hour
                timeslot = time.time() // 3600
                pipe.zincrby('tophits.%s' % timeslot, url, 1)

                # track slowest hits per hour (in ms)
                slowkey = 'elapsed.%s' % timeslot
                old_elapsed = r.zscore(slowkey, url)
                if old_elapsed is None or float(elapsed) > float(old_elapsed):
                    pipe.zincrby(slowkey, url, elapsed)

            pipe.execute()
    except (KeyboardInterrupt, SystemExit):
        pass
    except:
        raise
    finally:
        sub.close()

# What data do I want to fetch?
# 1. What URL is each client serving right now -- set of client:URL
# 2. What are the most popular URLs? -- sorted set of URL scored by weight
# 3. What are the most popular URLs within the past hour? -- sorted set of URLs scored by weight for each hour + cron job that updates a union of the past N buckets
# 4. What are the slowest URLs? -- sorted set of URLs scored by request time


collective/firehose/zope2.py
import time
import threading
import os
import redis
import zmq
from App.config import getConfiguration
from Products.Five import BrowserView

STATS = threading.local()

zmq_context = zmq.Context()
zmq_pub = zmq_context.socket(zmq.PUB)
zmq_pub.connect("ipc:///tmp/collective.firehose.sock")


config = getConfiguration()
try:
    instance_id = config.product_config['firehose']['instance_id']
except KeyError:
    instance_id = os.getpid()


def handle_start(event):
    request = event.request
    STATS.start = time.time()
    # XXX should include instance id
    zmq_pub.send('%s %s %s' % (instance_id, request.base + request.PATH_INFO, 0))


def handle_end(event):
    request = event.request
    zmq_pub.send('%s %s %s' % (instance_id, request.base + request.PATH_INFO, time.time() - STATS.start))


class StatsView(BrowserView):

    def update(self):
        r = redis.StrictRedis(host='localhost', port=6379, db=0)
        pipe = r.pipeline()

        timeslot = time.time() // 3600
        pipe.smembers('serving')
        pipe.zrevrange('tophits.%s' % timeslot, 0, 10, withscores=True, score_cast_func=int)
        pipe.zrevrange('elapsed.%s' % timeslot, 0, 10, withscores=True)

        self.serving, self.tophits, self.elapsed = pipe.execute()

    def __call__(self):
        self.update()
        return self.index()


# What data do I want to fetch?
# 1. What URL is each client serving right now -- set of client:URL
# 2. What are the most popular URLs? -- sorted set of URL scored by weight
# 3. What are the most popular URLs within the past hour? -- sorted set of URLs scored by weight for each hour + cron job that updates a union of the past N buckets
# 4. What are the slowest URLs? -- sorted set of URLs scored by request time
